File: problem17.php
<?php
//solution: 21124
echo eulerProblem17() . "\n";

function eulerProblem17() {
    $nums = array();
    $nums[] = "";
    $nums[] = "one";
    $nums[] = "two";
    $nums[] = "three";
    $nums[] = "four";
    $nums[] = "five";
    $nums[] = "six";
    $nums[] = "seven";
    $nums[] = "eight";
    $nums[] = "nine";
    $nums[] = "ten";
    $nums[] = "eleven";
    $nums[] = "twelve";
    $nums[] = "thirteen";
    $nums[] = "fourteen";
    $nums[] = "fifteen";
    $nums[] = "sixteen";
    $nums[] = "seventeen";
    $nums[] = "eighteen";
    $nums[] = "nineteen";
    
    $byten[] = "";
    $byten[] = "ten";
    $byten[] = "twenty";
    $byten[] = "thirty";
    $byten[] = "forty";
    $byten[] = "fifty";
    $byten[] = "sixty";
    $byten[] = "seventy";
    $byten[] = "eighty";
    $byten[] = "ninety";
    
    $sum = 0;
    for($i = 0; $i < count($nums); ++$i) {
        $sum += strlen($nums[$i]);
    }
    
    for($i = 20; $i < 1000; ++$i) {
        $f = intval(floor($i / 100));
        $s = intval(floor(($i - 100 * $f) / 10));
        $t = $i - 100 * $f - 10 * $s;
        
        if($f > 0) {
            $sum += strlen($nums[$f]."hundred");
        }
        
        if($s != 0 || $t != 0) {
            if($f > 0) {
                $sum += strlen("and");
            }
            
            if($s * 10 >= 20) {
                $sum += strlen($byten[$s]);
                $sum += strlen($nums[$t]);
            } else {
                $sum += strlen($nums[$s*10+$t]);
            }
        }
    }
    
    $sum += strlen("onethousand");
    
    return $sum;
}

?>

File: python/problem17.py
#!/usr/bin/python
#solution: 21124

def word_form(number):
    if number < 20:
        words = {
            0: "",
            1: "one",
            2: "two",
            3: "three",
            4: "four",
            5: "five",
            6: "six",
            7: "seven",
            8: "eight",
            9: "nine",
            10: "ten",
            11: "eleven",
            12: "twelve",
            13: "thirteen",
            14: "fourteen",
            15: "fifteen",
            16: "sixteen",
            17: "seventeen",
            18: "eighteen",
            19: "nineteen",
        }
        return words[number]

    if number < 100:
        tens = (number / 10) * 10
        ones = number % 10
        words = {
            20: "twenty",
            30: "thirty",
            40: "forty",
            50: "fifty",
            60: "sixty",
            70: "seventy",
            80: "eighty",
            90: "ninety"
        }
        return words[tens] + word_form(ones)

    if number < 1000:
        huns = (number / 100) * 100
        tens = ((number - huns) / 10) * 10
        ones = number - huns - tens
        sand = "and"
        if tens == 0 and ones == 0:
            sand = ""
        return word_form(huns / 100) + "hundred" + sand + word_form(tens + ones)

    return "onethousand"

def euler_problem_17():
    return len("".join([word_form(i) for i in range(1001)]))

print euler_problem_17()
